lib/compilers/d8.ts

```diff
@@ -0,0 +1,124 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+import fs from 'fs';
+import path from 'path';
+
+import {ParseFiltersAndOutputOptions} from '../../types/features/filters.interfaces';
+import {BaseCompiler} from '../base-compiler';
+import * as exec from '../exec';
+
+export class JavascriptCompiler extends BaseCompiler {
+    static get key() {
+        return 'd8';
+    }
+
+    override getIrOutputFilename(inputFilename: string, filters: ParseFiltersAndOutputOptions): string {
+        return this.filename(path.dirname(inputFilename) + '/code.asm');
+    }
+
+    public override getOutputFilename(dirPath: string, outputFilebase: string, key?: any) {
+        let filename;
+        if (key && key.backendOptions && key.backendOptions.customOutputFilename) {
+            filename = key.backendOptions.customOutputFilename;
+        } else {
+            filename = 'code.asm';
+        }
+
+        if (dirPath) {
+            return path.join(dirPath, filename);
+        } else {
+            return filename;
+        }
+    }
+
+    override optionsForFilter(filters: ParseFiltersAndOutputOptions, outputFilename: string, userOptions?: string[]) {
+        return [];
+    }
+
+    override async doCompilation(inputFilename, dirPath, key, options, filters, backendOptions, libraries, tools) {
+        const fileContents = await new Promise<string>((resolve, reject) => {
+            fs.readFile(inputFilename, 'utf8', (err, data) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(data);
+                }
+            });
+        });
+
+        const cppFileIdentifierMatch = fileContents.match(/\/\*\s*__(.*)_wasm__\s*\*\//);
+
+        if (cppFileIdentifierMatch) {
+            const cppFileName = cppFileIdentifierMatch[1];
+            const cppFilePath = path.join(dirPath, cppFileName);
+            const wasmFilePath = path.join(dirPath, 'out.wasm');
+            // compile cpp file to wasm
+            await exec.execute(
+                'clang',
+                [
+                    '-O2',
+                    '--target=wasm32',
+                    '--no-standard-libraries',
+                    '-Wl,--export-all',
+                    '-Wl,--no-entry',
+                    '-Wl,--import-memory',
+                    '-o',
+                    wasmFilePath,
+                    cppFilePath,
+                ],
+                {},
+            );
+            // read wasm file's bytes
+            const wasmBytes = await new Promise<Buffer>((resolve, reject) => {
+                fs.readFile(wasmFilePath, (err, wasmBytes) => {
+                    if (err) {
+                        reject(err);
+                    } else {
+                        resolve(wasmBytes);
+                    }
+                });
+            });
+
+            // Replace the bytes into the js file source
+            const wasmBytesAsString: string[] = Array.from({length: wasmBytes.length});
```

Comment:
can you refactor this into a separate function and call that

Reply:
this is also irrelevant now since I figured out how to not have to read the bytes inside the javascript compiler at all.